OpenRA.Mods.Common/AI/HackyAI.cs

```diff
@@ -559,6 +565,36 @@ public void Tick(Actor self)
 				World.IssueOrder(orders.Dequeue());
 		}
 
+		void ChooseEnemy()
+		{
+			if (Player.WinState != WinState.Undefined)
+				return;
+
+			var deadEnemies = World.Players
+				.Where(p => Player != p && Player.Stances[p] == Stance.Enemy && p.WinState == WinState.Lost);
+
+			var liveEnemies = World.Players
+				.Where(p => Player != p && !p.NonCombatant && Player.Stances[p] == Stance.Enemy && p.WinState == WinState.Undefined);
+
+			if (!liveEnemies.Any())
+				return;
+
+			// We don't want the AI to "attack" a player/AI that is already beaten
+			foreach (var d in deadEnemies)
+				aggro[d].Aggro = 0;
+
+			var leastLikedEnemies = liveEnemies
+				.GroupBy(e => aggro[e].Aggro)
+				.MaxByOrDefault(g => g.Key);
+
+			var enemy = (leastLikedEnemies != null) ?
+				leastLikedEnemies.Random(Random) : liveEnemies.FirstOrDefault();
+
+			// Bump the aggro slightly to avoid changing our mind
+			if (leastLikedEnemies.Count() > 1)
```

Comment:
I'm not sure whether there are any circumstances under which `leastLikedEnemies` can be `null` as long as there are still `liveEnemies`. I assumed there aren't any, but I'm not 100% sure how `MaxByOrDefault` works.

But yes, if it's possible for `leastLikedEnemies` to be null, that case needs to be covered, and if not, the `var enemy` can be simplified.